### ISSUE TITLE
Timezone issues in all-day events

### ISSUE THREAD

**Issue by wom-bat** [C4ptainCrunch/ics.py#92]
The ending date of all day events is off-by-one.

See attached test file.  I'd expect the output to be the same as the input.
[atest.py.txt](https://github.com/C4ptainCrunch/ics.py/files/509912/atest.py.txt)

**Comment by Tahlor** [comment]
I don't know if this is related, but I have a problem with the package explicitly setting the default time/timezone when parsing e.g. all-day events.

E.g.  if an input event had a start time like this: 
DTSTART;VALUE=DATE:20171202
When I print it out, it will look like this:
DTSTART:20171202T000000Z

**Comment by perette** [comment]
Looking at the .ics output, I'm seeing DTSTART for an `all_day()` event encoded correctly (;VALUE=DATE) but DTEND is still a specific time.  I'm also encountering DTEND missing entirely, which (although it's a "may" in RFC5545) is buggering up downstream utilities that expect it to be there.  Missingness seems a side-effect of `make_all_day()`, which sets end time to None if begin and end are the same time.  Since for an all-day event, a duration is implied, it seems outputting it would be best including DTEND even when neither end-time nor event-duration were explicitly set.

With that in mind, I propose these changes in event.py:
```
@Event._outputs
def o_all_day(event, container):
    if event.begin and event.all_day:
        container.append(ContentLine('DTSTART', params={'VALUE': ('DATE',)},
                                     value=arrow_date_to_iso(event.begin)))
+        if not event._duration:
+            container.append(ContentLine('DTEND', params={'VALUE': ('DATE',)},
+                                         value=arrow_date_to_iso(event.end)))
```
This adds DTEND encoding for all-day events to `o_all_day`, to match the way DTSTART is handled in the code for all-day events.  The `.end()` getter already makes the event 1 day long, if duration wasn't explicitly set.  Additionally, to avoid writing it twice, disable DTEND output in the regular handler when it's not all day:
```
@Event._outputs
def o_end(event, container):
-    if event.begin and event._end_time:
+    if event.begin and event._end_time and not event.all_day:
        container.append(ContentLine('DTEND', value=arrow_to_iso(event.end)))
```

**Comment by aureooms** [comment]
See https://github.com/C4ptainCrunch/ics.py/issues/150#issuecomment-429331431 for the possible source of this bug.

**Comment by C4ptainCrunch** [comment]
PR #144 might be of interest here

**Comment by C4ptainCrunch** [comment]
This minimal example reproduces the error
```python
from ics import Calendar
import arrow

cal = """
BEGIN:VCALENDAR
PRODID:-

BEGIN:VEVENT
DTSTART;VALUE=DATE:20161004
DTEND;VALUE=DATE:20161005
SUMMARY:An all day event: October 4 2016.
END:VEVENT

END:VCALENDAR
"""

c = Calendar(cal)
e = list(c.events)[0]

# This succeeds
assert e.begin == arrow.get('2016-10-04')
# This fails (because e.end = 2016-10-06)
assert e.end == arrow.get('2016-10-05')
```

**Comment by aureooms** [comment]
Just to make sure there is no misunderstanding: the start and end dates of the event are inclusive according to the RFC. Your example is really a two-day event happening on the 4th AND the 5th. It does not matter how this is handled internally (inclusive or exclusive bounds) as long as the result follows the RFC. My personal opinion is that the client API should at least allow to access this source data (without parsing the dump).

**Comment by aureooms** [comment]
Yes some of the test cases are wrong.

**Comment by adam-deutsch** [comment]
> Just to make sure there is no misunderstanding: the start and end dates of the event are inclusive according to the RFC. Your example is really a two-day event happening on the 4th AND the 5th. It does not matter how this is handled internally (inclusive or exclusive bounds) as long as the result follows the RFC. My personal opinion is that the client API should at least allow to access this source data (without parsing the dump).

Are you referring to RFC 5545? In section 3.6.1 I find the following text:

_The "DTSTART" property for a "VEVENT" specifies the inclusive start of the event.  For recurring events, it also specifies the very first instance in the recurrence set.  The "DTEND" property for a "VEVENT" calendar component specifies the non-inclusive end of the event._

In other words, DTSTART is indeed inclusive, but DTEND is exclusive.

**Comment by aureooms** [comment]
Let me check

**Comment by aureooms** [comment]
What I said above is wrong. Sorry for the confusion. I am indeed referring to that RFC and I am contradicting myself here https://github.com/C4ptainCrunch/ics.py/commit/bb3b3486aea6ce4d4e2478b3a42f43cebc40478a#diff-8c6cfa1a722bbdc29d84a31bd5e85c89R268.
Will check twice next time. `DTEND` is ***non***-inclusive.

**Comment by perette** [comment]
It appears the library API in the repository has moved to using inclusive dates for all-day events (i.e. New Years is a one-day event that starts January 1 and ends January 1).  This has broken the duration calculation, which now requires fudging to account for the additional day in all-day events.  It's also an API change, and has broken existing code in projects.

However, before fixing the duration, I suggest the inclusive-solution is misguided.  Experienced programmers use half-open ranges (i.e, New Years is a one-day event that starts January 1 and ends January 2) because they produce clean code without lots of fudges.  There may be other lurking edge cases introduced by moving to inclusive ranges.

I understand the users think in inclusive ranges, but the conversion is better handled at the user presentation layer.  And while making the API change might remove something that seems like a bug to newbies, [teaching why half-open ranges are good](https://stackoverflow.com/questions/13066884/what-is-half-open-range-and-off-the-end-value) is probably the better solution.

I am in favor of reverting this change.

**Comment by snoopyjc** [comment]
Yes, I'm having this problem too - just wasted a couple hours in the debugger trying to figure out why the 4th of July goes until the 6th!

**Comment by snoopyjc** [comment]
Timeline also says that the 4th of July Holiday is also an event on July 5th.  Errors in {included, overlapping, at} where you check start <= event.end instead of start < event.end

**Comment by 0187773933** [comment]
I think the arrow module reads the date as UTC timezone by default.
You have to manually set the timezone, and then it sets the right times.

```
from ics import Calendar, Event
import arrow

c = Calendar()
e = Event()
e.begin = arrow.get( "2020-09-06 00:00:01" , "YYYY-MM-DD HH:mm:ss" ).replace( tzinfo="US/Eastern" )
e.end = arrow.get( "2020-09-06 23:59:59" , "YYYY-MM-DD HH:mm:ss" ).replace( tzinfo="US/Eastern" )
e.name = "Test Event"
e.description = "Test Description"
c.events.add( e )
```
